### PR TITLE
fix: add dialog box for when user adds duplicate local electron builds

### PR DESCRIPTION
**Pull request by georgexu99**
Co-authored-by: VerteDinde <vertedinde@electronjs.org>

Users can be confused when adding a local electron build using the same folder path as a previously added (but not removed) build.

current behaviour checks that the path has been added before and ignores it, without feedback to the user.

This fix aims to resolve any user confusion as to why their local electron build "did not upload"

- [x] added check for existing local version
- [x] added snapshot test to ensure dialog box shows when expected

image

### CODE REVIEW

File: src/renderer/components/dialog-add-version.tsx [hunk 1]
```diff
@@ -168,30 +190,44 @@ export class AddVersionDialog extends React.Component<
...
+
   private renderVersionInput(): JSX.Element | null {
-    const { isValidElectron, isValidVersion, version } = this.state;
+    const {
+      existingLocalVersion,
+      isValidElectron,
+      isValidVersion,
+      version,
+    } = this.state;
+    const shouldRemovePreviousVersion = isValidElectron && existingLocalVersion;
```
Comment (erickzhao) [inline src/renderer/components/dialog-add-version.tsx:223]: I don't think checking for `isValidElectron` is necessary here because if this variable is false, we're already returning `null` right below.

**Review by erickzhao — CHANGES_REQUESTED** [review]
Seems like the renaming is a no-op for me when the local build is the active version. 🤔 

It might be easier to just select the existing version rather than performing the rename operation as well.

File: src/renderer/versions.ts [hunk 2]
```diff
@@ -179,6 +179,26 @@ export function addLocalVersion(input: Version): Array<Version> {
...
+  folderPath: string,
+): Version | undefined {
+  const versions = getLocalVersions();
+  let duplicate;
+
+  versions.find((v) => {
+    if (v.localPath === folderPath) duplicate = v;
+  });
+
+  return duplicate ? (duplicate as Version) : undefined;
```
[the remark this answers is not in the document]
Comment (georgexu99) [reply src/renderer/versions.ts:199]: oh! Embarassing :P Thanks Erick

File: src/renderer/versions.ts [hunk 3]
```diff
@@ -179,6 +179,22 @@ export function addLocalVersion(input: Version): Array<Version> {
...
+ * @param {Version} input
+ * @returns {Version | undefined}
+ */
+export function getExistingLocalVersion(
+  folderPath: string,
+): Version | undefined {
+  const versions = getLocalVersions();
+  const duplicate = versions.find((v) => v.localPath === folderPath);
+
+  return duplicate ? (duplicate as Version) : undefined;
```
Comment (ckerr) [inline src/renderer/versions.ts:195]: Array.prototype.find() returns the item or undefined, so I think we should be able to return it as-is without having to check the results:

```suggestion
  return getLocalVersions().find((v) => v.localPath === folderPath);
```

File: src/renderer/components/dialog-add-version.tsx [hunk 4]
```diff
@@ -168,25 +195,33 @@ export class AddVersionDialog extends React.Component<
...
       <Callout>
-        {info}
-        {this.renderVersionInput()}
+        {this.getVersionFromPathDialog()}
+        {!shouldSetToPreviousVersion && this.renderVersionInput()}
       </Callout>
     );
   }
 
+  private getVersionFromPathDialog(): string {
```
Comment (ckerr) [inline src/renderer/components/dialog-add-version.tsx:210]: This function name is slightly misleading; it implies a version will be returned but it's a message.

Maybe something like `getConfirmationMessage()` or `buildDialogText()`?

File: src/renderer/versions.ts [hunk 5]
```diff
@@ -179,6 +179,22 @@ export function addLocalVersion(input: Version): Array<Version> {
...
 }
 
+/**
+ * Check to see if local version has already been added
+ * for the current folder path
+ *
+ * @param {Version} input
+ * @returns {Version | undefined}
+ */
+export function getExistingLocalVersion(
```
Comment (ckerr) [inline src/renderer/versions.ts:189]: Another naming suggestion for consistency with other utils like `getElectronNameForPlatform()`, `getLocalTypePathForVersion()`, etc

```suggestion
export function getLocalVersionForPath(
```

File: src/renderer/versions.ts [hunk 6]
```diff
@@ -179,6 +179,22 @@ export function addLocalVersion(input: Version): Array<Version> {
   return versions;
 }
 
+/**
+ * Check to see if local version has already been added
+ * for the current folder path
+ *
+ * @param {Version} input
```
Comment (ckerr) [inline src/renderer/versions.ts:186]:
```suggestion
 * @param {string} folderPath
```

File: src/renderer/versions.ts [hunk 7]
```diff
@@ -179,6 +179,22 @@ export function addLocalVersion(input: Version): Array<Version> {
   return versions;
 }
 
+/**
+ * Check to see if local version has already been added
+ * for the current folder path
```
Comment (ckerr) [inline src/renderer/versions.ts:184]:
```suggestion
 * Get the Version (if any) that is located at localPath.
```

File: src/renderer/components/dialog-add-version.tsx [hunk 8]
```diff
@@ -103,7 +112,18 @@ export class AddVersionDialog extends React.Component<
...
-    this.props.appState.addLocalVersion(toAdd);
+    // swap to old local electron version if the user adds a new one with the same path
+    const shouldSetToPreviousVersion = isValidElectron && existingLocalVersion;
+    if (shouldSetToPreviousVersion) {
+      // set previous version as active version
+      for (const ver of Object.values(versions)) {
+        if (ver.localPath === existingLocalVersion?.localPath) {
+          this.props.appState.setVersion(ver.version);
+        }
+      }
```
Comment (ckerr) [inline src/renderer/components/dialog-add-version.tsx:123]: Could this for loop be replaced with `this.props.appState.setVersion(existingLocalVersion.localPath)`?

It looks like there's something going on here that makes walking the `appState.versions` array necessary, but I'm not seeing why that's needed instead of using `existingLocalVersion.localPath` :thinking:

File: src/renderer/components/dialog-add-version.tsx [hunk 9]
```diff
@@ -103,7 +112,18 @@ export class AddVersionDialog extends React.Component<
       name,
     };
 
-    this.props.appState.addLocalVersion(toAdd);
+    // swap to old local electron version if the user adds a new one with the same path
+    const shouldSetToPreviousVersion = isValidElectron && existingLocalVersion;
+    if (shouldSetToPreviousVersion) {
+      // set previous version as active version
+      for (const ver of Object.values(versions)) {
+        if (ver.localPath === existingLocalVersion?.localPath) {
```
Comment (ckerr) [inline src/renderer/components/dialog-add-version.tsx:120]: `?` is not needed; we know from line 116 that `existingLocalVersion` is truthy

File: src/renderer/components/dialog-add-version.tsx [hunk 10]
```diff
@@ -116,15 +136,22 @@ export class AddVersionDialog extends React.Component<
...
+    const canSubmit = isValidElectron && isValidVersion;
+    // swap to old local electron version if the user adds a new one with the same path
+    const shouldSetToPreviousVersion = isValidElectron && existingLocalVersion;
 
     return [
       <Button
         icon="add"
         key="submit"
-        disabled={!canSubmit}
+        disabled={shouldSetToPreviousVersion ? false : !canSubmit}
```
Comment (ckerr) [inline src/renderer/components/dialog-add-version.tsx:152]:
```suggestion
        disabled={!shouldSetToPreviousVersion && !canSubmit}
```

File: src/renderer/components/dialog-add-version.tsx (hunk 8 above)
Comment (georgexu99) [reply src/renderer/components/dialog-add-version.tsx:123]: Oops! You are totally right- an oversight for when line 121 used to delete the previous version in a previous commit that I ended up changing 😆

File: src/renderer/components/dialog-add-version.tsx [hunk 11]
```diff
@@ -168,25 +190,33 @@ export class AddVersionDialog extends React.Component<
...
       </Callout>
     );
   }
 
+  private buildDialogText(): string {
+    const { existingLocalVersion, isValidElectron } = this.state;
+
+    if (isValidElectron && existingLocalVersion) {
+      return `This folder path has already been added as version "${existingLocalVersion.version}".
+      Would you like to set your Electron version to "${existingLocalVersion.version}"? \n \n`;
```
Comment (ckerr) [inline src/renderer/components/dialog-add-version.tsx:210]:
```suggestion
      return `This folder is already in use as version "${existingLocalVersion.version}". Would you like to switch to that version now?`;
```

File: src/renderer/components/dialog-add-version.tsx [hunk 12]
```diff
@@ -116,15 +131,22 @@ export class AddVersionDialog extends React.Component<
...
   get buttons() {
-    const canSubmit = this.state.isValidElectron && this.state.isValidVersion;
+    const {
+      isValidElectron,
+      isValidVersion,
+      existingLocalVersion,
+    } = this.state;
+    const canSubmit = isValidElectron && isValidVersion;
+    // swap to old local electron version if the user adds a new one with the same path
+    const shouldSetToPreviousVersion = isValidElectron && existingLocalVersion;
```
Comment (ckerr) [inline src/renderer/components/dialog-add-version.tsx:141]: I know I had a run at reviewing this section before with the `disabled={foo}` line but this whole section is still hard for me to scan, so I'm going to try meddling again. :slightly_smiling_face: 

- I think part of it is that we can still submit even when `canSubmit` is false since this dialog now has a dual role.
- Another part is that `shouldSetToPreviousVersion` is a little unclear, we could rename it to draw a line between it and `existingLocalVersion` more clearly.
- Thirdly, the `disabled={foo}` logic, even revised, is still kind of hard to scan.

```suggestion
    const canAdd = isValidElectron && isValidVersion && !existingLocalVersion;
    const canSwitch = isValidElectron && existingLocalVersion;
```

To me this is a lot easier to read, e.g. `canAdd` and `canSwitch` are terse and their relationship to each other is easy to understand at-a-glance. And this also makes the `disabled={}` line clearer: `disabled={!canAdd && !canSwitch}`

File: src/renderer/components/dialog-add-version.tsx [hunk 13]
```diff
@@ -116,15 +131,22 @@ export class AddVersionDialog extends React.Component<
...
 
     return [
       <Button
         icon="add"
         key="submit"
-        disabled={!canSubmit}
+        disabled={!shouldSetToPreviousVersion && !canSubmit}
```
Comment (ckerr) [inline src/renderer/components/dialog-add-version.tsx:147]:
```suggestion
        disabled={!canAdd && !canSwitch}
```

File: src/renderer/components/dialog-add-version.tsx [hunk 14]
```diff
@@ -116,15 +131,22 @@ export class AddVersionDialog extends React.Component<
...
         onClick={this.onSubmit}
-        text="Add"
+        text={shouldSetToPreviousVersion ? 'OK' : 'Add'}
```
Comment (ckerr) [inline src/renderer/components/dialog-add-version.tsx:149]:
```suggestion
        text={canSwitch ? 'Switch' : 'Add'}
```

**Review by ckerr — COMMENTED** [review]
I had a couple more suggestions but nothing structural. Overall, this is looking good!

File: src/renderer/components/dialog-add-version.tsx [hunk 15]
```diff
@@ -168,25 +190,33 @@ export class AddVersionDialog extends React.Component<
   }
 
   private renderPath(): JSX.Element | null {
-    const { folderPath, isValidElectron } = this.state;
+    const { isValidElectron, folderPath, existingLocalVersion } = this.state;
+    const shouldSetToPreviousVersion = isValidElectron && existingLocalVersion;
```
Comment (ckerr) [inline src/renderer/components/dialog-add-version.tsx:194]:
```suggestion
    const canSwitch = isValidElectron && existingLocalVersion;
```

For consistency with other suggested change

File: src/renderer/components/dialog-add-version.tsx [hunk 16]
```diff
@@ -168,25 +190,33 @@ export class AddVersionDialog extends React.Component<
...
+        {this.buildDialogText()}
+        {!shouldSetToPreviousVersion && this.renderVersionInput()}
       </Callout>
     );
   }
 
+  private buildDialogText(): string {
+    const { existingLocalVersion, isValidElectron } = this.state;
+
+    if (isValidElectron && existingLocalVersion) {
```
Comment (ckerr) [inline src/renderer/components/dialog-add-version.tsx:208]: Since we've got the same of (isValidElectron && existingLocalVesion) logic repeated here again, I'd probably use the same variables here too for clarity's sake, e.g.

```
const canAdd = isValidElectron && isValidVersion && !existingLocalVersion;
const canSwitch = isValidElectron && existingLocalVersion;

if (canSwitch) {
  return ...
}
if (canAdd) {
  return ...
}
return ...
```

Comment (georgexu99) [reply src/renderer/components/dialog-add-version.tsx:208]: I apologize if I'm mistaken, but I took a look at this with @VerteDinde for a bit on Friday and to my understanding, we were unable to make variable names work like you described here due to some typescript issues like so: 

image

Comment (ckerr) [reply src/renderer/components/dialog-add-version.tsx:209]: @georgexu99 this is more a nice-to-have and not a blocker.

However, in that screenshot I think at line 211 we know that `existingLocalVersion` is defined because of the test on line 209. So we just have to explain that to TypeScript by appending a `!` to the variable name e.g.

```diff
- ${existingLocalVersion.version}
+ ${existingLocalVersion!.version}
```

File: src/renderer/components/dialog-add-version.tsx [hunk 17]
```diff
@@ -116,15 +131,23 @@ export class AddVersionDialog extends React.Component<
...
 
     return [
       <Button
         icon="add"
         key="submit"
-        disabled={!canSubmit}
+        disabled={!canAdd && !canSwitch}
         onClick={this.onSubmit}
-        text="Add"
+        text={shouldSetToPreviousVersion ? 'Switch' : 'Add'}
```
Comment (ckerr) [inline src/renderer/components/dialog-add-version.tsx:150]:
```suggestion
        text={canSwitch ? 'Switch' : 'Add'}
```

File: src/renderer/components/dialog-add-version.tsx [hunk 18]
```diff
@@ -116,15 +131,23 @@ export class AddVersionDialog extends React.Component<
...
-    const canSubmit = this.state.isValidElectron && this.state.isValidVersion;
+    const {
+      isValidElectron,
+      isValidVersion,
+      existingLocalVersion,
+    } = this.state;
+    const canAdd = isValidElectron && isValidVersion && !existingLocalVersion;
+    const canSwitch = isValidElectron && existingLocalVersion;
+    // swap to old local electron version if the user adds a new one with the same path
+    const shouldSetToPreviousVersion = isValidElectron && existingLocalVersion;
```
Comment (ckerr) [inline src/renderer/components/dialog-add-version.tsx:142]: `shouldSetToPreviousVersion` is a little prolix and its relationship to the other fields isn't clear at-a-glance. In particular someone who hasn't read this PR might not be able to tell if 'previous version' is the same thing as 'existing local version'.

It's also identical to the value of `canSwitch` so we probably don't need both.

There's real value in being able to understand code at-a-glance:

> `{text=canSwitch ? 'Switch' : 'Add'}`

Is something a developer can read and understand as soon as their eyes see it. It's short, it's direct, and it follows the hints the rest of the code has with use of the word 'switch', e.g. it's easy to understand the relationship between the two uses of 'switch' in that line.

> `shouldSetToPreviousVersion`

is slower to read because it takes longer to parse the words in the name and understand how they relate to the rest of the code.

```suggestion
```

**Review by ckerr — COMMENTED** [review]
Thanks for the changes!

File: src/renderer/components/dialog-add-version.tsx [hunk 19]
```diff
@@ -168,25 +191,32 @@ export class AddVersionDialog extends React.Component<
...
+    const { existingLocalVersion, isValidElectron } = this.state;
+
+    if (isValidElectron && existingLocalVersion) {
+      return `This folder is already in use as version "${existingLocalVersion.version}". Would you like to switch to that version now?`;
+    } else if (isValidElectron) {
+      return `We found an ${getElectronNameForPlatform()} in this folder.`;
+    } else {
+      return `We did not find a ${getElectronNameForPlatform()} in this folder...`;
+    }
+  }
```
Comment (ckerr) [inline src/renderer/components/dialog-add-version.tsx:216]:
```suggestion
  private buildDialogText(): string { 
    const { isValidElectron, isValidVersion, existingLocalVersion } = this.state;
    const canAdd = isValidElectron && isValidVersion && !existingLocalVersion;
    const canSwitch = isValidElectron && existingLocalVersion;

    if (canSwitch)
      return `This folder is already in use as version "${existingLocalVersion!.version}". Would you like to switch to that version now?`;

    if (canAdd)
      return `We found an ${getElectronNameForPlatform()} in this folder.`;

    return `We did not find a ${getElectronNameForPlatform()} in this folder...`;
  } 

```

File: src/renderer/components/dialog-add-version.tsx (hunk 16 above)
Comment (georgexu99) [reply src/renderer/components/dialog-add-version.tsx:209]: oh Thank you!

File: src/renderer/components/dialog-add-version.tsx (hunk 18 above)
Comment (georgexu99) [reply src/renderer/components/dialog-add-version.tsx:142]: Thanks for this feedback, I'll keep this in mind for the future <3

File: src/renderer/components/dialog-add-version.tsx (hunk 19 above)
Comment (georgexu99) [reply src/renderer/components/dialog-add-version.tsx]: The second block on line 214, i think the conditional must stay as Felix's old implementation 'isValidElectron'. If we change it as 'canAdd' and add 'isValidVersion' and '!existingLocalVersion' , the logic changes and we don't get the text we expect in certain test cases

I will however change the if-else styling to remove the curly brackets (is this a style choice or is it objectively better?)

Comment (ckerr) [reply src/renderer/components/dialog-add-version.tsx]: Yes, you're right about the logic. Good catch.

No, the if-else braces are subjective and you can do whatever you like. Now that you mention it, the style I used in my suggested change is probably subjectively worse since it's not as consistent with the rest of the code around it, so probably better to keep the braces.

It's arguably better to not use 'else' after 'return' since it's an unnecessary code instruction that the compiler will have to parse and then discard. How much difference that overhead makes in the real world is probably negligible, but that's the closest I know of to an _objective_ answer of what would be the better way to write that function.

tldr; probably keep the braces in, and keep the elses out.

Anyway, thanks for your patience with all the variable & function naming. :slightly_smiling_face:

**Review by ckerr — APPROVED** [review]

**Review by MobarakeGH — CHANGES_REQUESTED** [review]
#744

**Review by erickzhao — APPROVED** [review]
